csharp/src/Apache.Arrow/C/CArrowArrayExporter.cs

```diff
@@ -0,0 +1,210 @@
+                cArray->release = (delegate* unmanaged[Stdcall]<CArrowArray*, void>)Marshal.GetFunctionPointerForDelegate<ReleaseArrowArray>(ReleaseArray);
+                cArray->private_data = FromDisposable(allocationOwner);
+                allocationOwner = null;
+            }
+            finally
+            {
+                allocationOwner?.Dispose();
+            }
+        }
+
+        private unsafe static void ConvertArray(ExportedAllocationOwner sharedOwner, ArrayData array, CArrowArray* cArray)
+        {
+            cArray->length = array.Length;
+            cArray->offset = array.Offset;
+            cArray->null_count = array.NullCount;
+            cArray->release = (delegate* unmanaged[Stdcall]<CArrowArray*, void>)Marshal.GetFunctionPointerForDelegate<ReleaseArrowArray>(ReleaseArray);
+            cArray->private_data = null;
+
+            cArray->n_buffers = array.Buffers?.Length ?? 0;
+            cArray->buffers = null;
+            if (cArray->n_buffers > 0)
+            {
+                cArray->buffers = (byte**)Marshal.AllocCoTaskMem(array.Buffers.Length * IntPtr.Size);
+                for (int i = 0; i < array.Buffers.Length; i++)
+                {
+                    ArrowBuffer buffer = array.Buffers[i];
+                    IntPtr ptr;
+                    if (!buffer.TryExport(sharedOwner, out ptr))
+                    {
+                        throw new NotSupportedException(); // TODO
+                    }
+                    cArray->buffers[i] = (byte*)ptr;
+                }
+            }
+
+            cArray->n_children = array.Children?.Length ?? 0;
+            cArray->children = null;
+            if (cArray->n_children > 0)
+            {
+                cArray->children = (CArrowArray**)Marshal.AllocCoTaskMem(IntPtr.Size * array.Children.Length);
+                for (int i = 0; i < array.Children.Length; i++)
+                {
+                    cArray->children[i] = CArrowArray.Create();
+                    ConvertArray(sharedOwner, array.Children[i], cArray->children[i]);
+                }
+            }
+
+            cArray->dictionary = null;
+            if (array.Dictionary != null)
+            {
+                cArray->dictionary = CArrowArray.Create();
+                ConvertArray(sharedOwner, array.Dictionary, cArray->dictionary);
+            }
+        }
+
+        private unsafe static void ConvertRecordBatch(ExportedAllocationOwner sharedOwner, RecordBatch batch, CArrowArray* cArray)
+        {
+            cArray->length = batch.Length;
+            cArray->offset = 0;
+            cArray->null_count = 0;
+            cArray->release = (delegate* unmanaged[Stdcall]<CArrowArray*, void>)Marshal.GetFunctionPointerForDelegate<ReleaseArrowArray>(ReleaseArray);
+            cArray->private_data = null;
+
+            cArray->n_buffers = 1;
+            cArray->buffers = (byte**)Marshal.AllocCoTaskMem(IntPtr.Size);
```

Comment:
Why are we using `AllocCoTaskMem` here and in other places?

Reply:
derp